datastream-server/src/main/java/com/linkedin/datastream/server/CoordinatorEventBlockingQueue.java

```diff
@@ -73,16 +73,30 @@ class CoordinatorEventBlockingQueue implements MetricsAware {
 
 
   /**
-   * Add a single event to the queue, overwriting events with the same name and same metadata.
+   * Add a single event to the queue. Defaults to adding the event at the end of the queue.
    * @param event CoordinatorEvent event to add to the queue
    */
   public synchronized void put(CoordinatorEvent event) {
-    LOG.info("Queuing event {} to event queue", event.getType());
+    put(event, true);
+  }
+
+  /**
+   * Add a single event to the queue, de-duping events with the same name and same metadata.
+   * @param event CoordinatorEvent event to add to the queue
+   * @param insertInTheEnd if true, indicates to add the event to the end of the queue and front, otherwise.
+   */
+  public synchronized void put(CoordinatorEvent event, boolean insertInTheEnd) {
```

Comment:
for understanding: LinkedBlockingDeque is already a thread safe and we are using only offer or offerFirst, what is the rational on having this method synchronized?

Reply:
The synchronized methods were added long before but I am guessing that the shared set variable might create race and would let in duplicate events in some cases if those methods are not synchronized. 

That is why the wrapper on top of the LinkedBlockingDeque is implemented which only supports taking out events from the front.